internal/storage/sql/auth/store_test.go

```diff
@@ -0,0 +1,276 @@
+				Metadata: map[string]string{
+					"io.flipt.auth.token.name":        "access_all_areas",
+					"io.flipt.auth.token.description": "The keys to the castle",
+				},
+			},
+			expectedErrIs: storage.ErrInvalid,
+		},
+		{
+			name: "fails token uniqueness constraint",
+			opts: func(t *testing.T) []Option {
+				return []Option{
+					WithTokenGeneratorFunc(func() string {
+						// return previous tests created token
+						return "create_auth_token"
+					}),
+				}
+			},
+			req: &storage.CreateAuthenticationRequest{
+				Method:    auth.Method_TOKEN,
+				ExpiresAt: timestamppb.New(time.Unix(2, 0)),
+				Metadata: map[string]string{
+					"io.flipt.auth.token.name":        "access_all_areas",
+					"io.flipt.auth.token.description": "The keys to the castle",
+				},
+			},
+			expectedErrIs: storage.ErrInvalid,
+		},
+	} {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			store := storeFn(test.opts(t)...)
+
+			clientToken, created, err := store.CreateAuthentication(ctx, test.req)
+			if test.expectedErrIs != nil {
+				require.ErrorIs(t, err, test.expectedErrIs)
+				return
+			}
+
+			require.NoError(t, err)
+			assert.Equal(t, test.expectedToken, clientToken)
+			assert.Equal(t, test.expectedAuthentication, created)
+		})
+	}
+}
+
+func TestAuthentication_GetAuthenticationByClientToken(t *testing.T) {
+	// seed database state
+	ctx := context.TODO()
+
+	// established a store factory with a single seeded auth entry
+	storeFn := newTestStore(t, createAuth("get_auth_id", "get_auth_token"))
+
+	// run table tests
+	for _, test := range []struct {
+		name                   string
+		clientToken            string
+		expectedErrIs          error
+		expectedAuthentication *auth.Authentication
+	}{
+		{
+			name:          "error not found for unexpected clientToken",
+			clientToken:   "unknown",
+			expectedErrIs: storage.ErrNotFound,
+		},
+		{
+			name:        "successfully retrieves authentication by clientToken",
+			clientToken: "get_auth_token",
+			expectedAuthentication: &auth.Authentication{
+				Id:     "get_auth_id",
+				Method: auth.Method_TOKEN,
+				Metadata: map[string]string{
+					"io.flipt.auth.token.name":        "access_some_areas",
+					"io.flipt.auth.token.description": "The keys to some of the castle",
+				},
+				ExpiresAt: timestamppb.New(time.Unix(2, 0)),
+				CreatedAt: someTimestamp,
+				UpdatedAt: someTimestamp,
+			},
+		},
+	} {
+		var (
+			clientToken            = test.clientToken
+			expectedErrIs          = test.expectedErrIs
+			expectedAuthentication = test.expectedAuthentication
+		)
+
+		t.Run(test.name, func(t *testing.T) {
+			retrieved, err := storeFn(commonOpts(t)...).GetAuthenticationByClientToken(ctx, clientToken)
+			if expectedErrIs != nil {
+				require.ErrorIs(t, err, expectedErrIs)
+				return
+			}
+
+			require.NoError(t, err)
+			assert.Equal(t, expectedAuthentication, retrieved)
+		})
+	}
+}
+
+func FuzzHashClientToken(f *testing.F) {
```

Comment:
ooo love the fuzz

Reply:
Honestly, first time I have given it a proper go. It is really neat.
I ran a long fuzz of this function. Roughly ~7M calls.

It is mostly just base64 and sha256, so not testing much.
However, it was nice to get some confidence it was producing base64 encoded non-zero length tokens.